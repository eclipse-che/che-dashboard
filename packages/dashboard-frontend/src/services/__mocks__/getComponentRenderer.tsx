/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';

export default function <T extends Array<unknown>>(
  getComponent: (..._args: T) => React.ReactElement,
): {
  createSnapshot: (...args: T) => renderer.ReactTestRenderer;
  renderComponent: (...args: T) => {
    reRenderComponent: (..._args: T) => void;
  };
} {
  return {
    createSnapshot: (...args: T) => {
      return renderer.create(getComponent(...args));
    },
    renderComponent: (...args: T) => {
      const res = render(getComponent(...args));
      return {
        reRenderComponent: (..._args: T) => {
          res.rerender(getComponent(..._args));
        },
      };
    },
  };
}
