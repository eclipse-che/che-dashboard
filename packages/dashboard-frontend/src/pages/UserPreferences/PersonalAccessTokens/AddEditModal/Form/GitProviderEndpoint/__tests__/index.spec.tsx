/*
 * Copyright (c) 2018-2023 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { Form } from '@patternfly/react-core';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { GitProviderEndpoint } from '..';
import getComponentRenderer, {
  screen,
} from '../../../../../../../services/__mocks__/getComponentRenderer';

const { createSnapshot, renderComponent } = getComponentRenderer(getComponent);

const mockOnChange = jest.fn();

describe('GitProviderEndpoint', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test('snapshot w/o endpoint', () => {
    const snapshot = createSnapshot();
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  test('snapshot with endpoint', () => {
    const snapshot = createSnapshot('https://provider/endpoint');
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  it('should handle a correct endpoint', () => {
    const endpoint = 'https://provider/endpoint';
    renderComponent();

    expect(mockOnChange).not.toHaveBeenCalled();

    const input = screen.getByRole('textbox');
    userEvent.paste(input, endpoint);

    expect(mockOnChange).toHaveBeenCalledWith(endpoint, true);
    expect(screen.queryByText('The URL is not valid.')).toBeFalsy();
  });

  it('should handle endpoint started with an incorrect protocol', () => {
    const endpoint = 'asdf://provider/endpoint';
    renderComponent();

    expect(mockOnChange).not.toHaveBeenCalled();

    const input = screen.getByRole('textbox');
    userEvent.paste(input, endpoint);

    expect(mockOnChange).toHaveBeenCalledWith(endpoint, false);
    expect(screen.queryByText('The URL is not valid.')).toBeTruthy();
  });

  it('should handle endpoint w/o protocol', () => {
    const endpoint = 'provider/endpoint';
    renderComponent();

    expect(mockOnChange).not.toHaveBeenCalled();

    const input = screen.getByRole('textbox');
    userEvent.paste(input, endpoint);

    expect(mockOnChange).toHaveBeenCalledWith(endpoint, false);
    expect(screen.queryByText('The URL is not valid.')).toBeTruthy();
  });

  it('should handle an empty value', () => {
    const endpoint = 'https://provider/endpoint';
    renderComponent(endpoint);

    expect(mockOnChange).not.toHaveBeenCalled();

    const input = screen.getByRole('textbox');
    userEvent.clear(input);

    expect(mockOnChange).toHaveBeenCalledWith('', false);
    expect(screen.queryByText('This field is required.')).toBeTruthy();
  });
});

function getComponent(providerEndpoint?: string): React.ReactElement {
  return (
    <Form>
      <GitProviderEndpoint
        providerEndpoint={providerEndpoint}
        onChange={(...args) => mockOnChange(...args)}
      />
    </Form>
  );
}
