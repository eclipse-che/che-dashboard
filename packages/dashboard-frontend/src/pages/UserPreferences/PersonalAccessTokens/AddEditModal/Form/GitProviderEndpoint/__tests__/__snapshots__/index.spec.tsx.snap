// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitProviderEndpoint snapshot w/o endpoint 1`] = `
<form
  className="pf-c-form"
  noValidate={true}
>
  <div
    className="pf-c-form__group"
  >
    <div
      className="pf-c-form__group-label"
    >
      <label
        className="pf-c-form__label"
        htmlFor="git-provider-endpoint-label"
      >
        <span
          className="pf-c-form__label-text"
        >
          Git Provider Endpoint
        </span>
        <span
          aria-hidden="true"
          className="pf-c-form__label-required"
        >
           
          *
        </span>
      </label>
       
    </div>
    <div
      className="pf-c-form__group-control"
    >
      <input
        aria-describedby="git-provider-endpoint-label"
        aria-invalid={false}
        aria-label="Git Provider Endpoint"
        className="pf-c-form-control"
        data-ouia-component-id="OUIA-Generated-TextInputBase-1"
        data-ouia-component-type="PF4/TextInput"
        data-ouia-safe={true}
        disabled={false}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Enter a Git Provider Endpoint"
        required={false}
        type="url"
        value="https://github.com"
      />
    </div>
  </div>
</form>
`;

exports[`GitProviderEndpoint snapshot with endpoint 1`] = `
<form
  className="pf-c-form"
  noValidate={true}
>
  <div
    className="pf-c-form__group"
  >
    <div
      className="pf-c-form__group-label"
    >
      <label
        className="pf-c-form__label"
        htmlFor="git-provider-endpoint-label"
      >
        <span
          className="pf-c-form__label-text"
        >
          Git Provider Endpoint
        </span>
        <span
          aria-hidden="true"
          className="pf-c-form__label-required"
        >
           
          *
        </span>
      </label>
       
    </div>
    <div
      className="pf-c-form__group-control"
    >
      <input
        aria-describedby="git-provider-endpoint-label"
        aria-invalid={false}
        aria-label="Git Provider Endpoint"
        className="pf-c-form-control"
        data-ouia-component-id="OUIA-Generated-TextInputBase-2"
        data-ouia-component-type="PF4/TextInput"
        data-ouia-safe={true}
        disabled={false}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Enter a Git Provider Endpoint"
        required={false}
        type="url"
        value="https://provider.test/endpoint"
      />
    </div>
  </div>
</form>
`;
